zebra-chain/src/orchard/commitment.rs

```diff
@@ -0,0 +1,425 @@
+    }
+}
+
+impl From<pallas::Point> for ValueCommitment {
+    fn from(projective_point: pallas::Point) -> Self {
+        Self(pallas::Affine::from(projective_point))
+    }
+}
+
+impl Eq for ValueCommitment {}
+
+/// LEBS2OSP256(repr_P(cv))
+///
+/// https://zips.z.cash/protocol/protocol.pdf#pallasandvesta
+impl From<ValueCommitment> for [u8; 32] {
+    fn from(cm: ValueCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+impl<'a> std::ops::Sub<&'a ValueCommitment> for ValueCommitment {
+    type Output = Self;
+
+    fn sub(self, rhs: &'a ValueCommitment) -> Self::Output {
+        self - *rhs
+    }
+}
+
+impl std::ops::Sub<ValueCommitment> for ValueCommitment {
+    type Output = Self;
+
+    fn sub(self, rhs: ValueCommitment) -> Self::Output {
+        ValueCommitment((self.0 - rhs.0).into())
+    }
+}
+
+impl std::ops::SubAssign<ValueCommitment> for ValueCommitment {
+    fn sub_assign(&mut self, rhs: ValueCommitment) {
+        *self = *self - rhs;
+    }
+}
+
+impl std::iter::Sum for ValueCommitment {
+    fn sum<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = Self>,
+    {
+        iter.fold(
+            ValueCommitment(pallas::Affine::identity()),
+            std::ops::Add::add,
+        )
+    }
+}
+
+/// LEBS2OSP256(repr_P(cv))
+///
+/// https://zips.z.cash/protocol/protocol.pdf#pallasandvesta
+impl TryFrom<[u8; 32]> for ValueCommitment {
+    type Error = &'static str;
+
+    fn try_from(bytes: [u8; 32]) -> Result<Self, Self::Error> {
+        let possible_point = pallas::Affine::from_bytes(&bytes);
+
+        if possible_point.is_some().into() {
+            Ok(Self(possible_point.unwrap()))
+        } else {
+            Err("Invalid pallas::Affine value")
+        }
+    }
+}
+
+impl ZcashSerialize for ValueCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&<[u8; 32]>::from(*self)[..])?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for ValueCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Self::try_from(reader.read_32_bytes()?).map_err(|e| SerializationError::Parse(e))
+    }
+}
+
+impl ValueCommitment {
+    /// Generate a new _ValueCommitment_.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+    pub fn randomized<T>(csprng: &mut T, value: Amount) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        let rcv = generate_trapdoor(csprng);
+
+        Self::new(rcv, value)
+    }
+
+    /// Generate a new _ValueCommitment_ from an existing _rcv_ on a _value_.
+    ///
+    /// ValueCommit^Orchard(v) :=
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+    #[allow(non_snake_case)]
+    pub fn new(rcv: pallas::Scalar, value: Amount) -> Self {
+        let V = pallas_group_hash(b"z.cash:Orchard-cv", b"v");
+        let R = pallas_group_hash(b"z.cash:Orchard-cv", b"r");
```

Comment:
The group hash is quite fast but it still requires computing a square root, which is about the cost of a full-width scalar mul. So that's about 3.25 full-width scalar muls to compute a value commitment instead of 1.25. Presumably the other commitments in Orchard will have the same issue, so it's not an insignificant cost.

Reply:
Ouse should lazy_static! these values as well.

Reply:
Done in b848cb3733f69f5e258b978d1433040c3ac0497d